lib/collection/src/shards/local_shard/scroll.rs

```diff
@@ -280,51 +281,57 @@ impl LocalShard {
+                "Failed to create weighted index for random scroll: {:?}",
+                err
+            ))
+        })?;
 
         // Get `limit` points in a weighted fashion from each segment, depending on how many points each segment has.
         let mut rng = rand::thread_rng();
-        let mut random_points = Vec::with_capacity(limit);
-        let mut exhausted_reads = HashSet::new();
-        for segment_id in (0..segments_reads.len()).cycle() {
-            if exhausted_reads.contains(&segment_id) {
-                continue;
-            }
-
-            // Determine weight for this segment
-            let weight = segment_weights.get(segment_id).unwrap_or(&0.5);
-
-            if !rng.gen_bool(*weight) {
-                continue;
+        let mut random_points = HashSet::with_capacity(limit);
+
+        // This loop iterates <= LIMIT times, and either quits if we have enough points,
+        // or if some of the segments are exhausted.
+        //
+        // If the segments are exhausted, we will fill up the rest of the points from other segments.
+        // In total, the complexity is guaranteed to be O(limit).
+        while random_points.len() < limit {
+            let segment_offset = rng.sample(&distribution);
+            let points = segments_reads.get_mut(segment_offset).unwrap();
+            if let Some(point) = points.pop() {
+                random_points.insert(point);
+            } else {
+                // It seems that some segments are empty early,
+                // so distribution does not make sense anymore.
+                // This is only possible if segments size < limit.
+                break;
             }
+        }
 
-            let Some(point) = segments_reads
-                .get_mut(segment_id)
-                .and_then(|points| points.pop())
-            else {
-                // reads have been exhausted, don't use this segment anymore
-                exhausted_reads.insert(segment_id);
-                if exhausted_reads.len() == segments_reads.len() {
-                    // no points left in any read
+        // If we still need more points, we will get them from the rest of the segments.
+        // This is a rare case, as it seems we don't have enough points in individual segments.
+        // Therefore, we can ignore "proper" distribution, as it won't be accurate anyway.
+        if random_points.len() < limit {
+            let rest_points = segments_reads.into_iter().flatten();
+            for point in rest_points {
+                random_points.insert(point);
+                if random_points.len() >= limit {
                     break;
                 }
-                continue;
-            };
-
-            // Add point to result
-            random_points.push(point);
-            if random_points.len() == limit {
-                break;
             }
         }
 
+        let selected_points: Vec<_> = random_points.into_iter().collect();
```

Comment:
Is it possible for the output to be somewhat deterministic here due to the hash set? I honestly don't know.

If it is, I think that's undesirable.

Reply:
Assuming each set has an unique [`RandomState`](https://doc.rust-lang.org/std/collections/hash_map/struct.RandomState.html), it does not have this problem.